x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptionPacketsInputStream.java

```diff
@@ -0,0 +1,160 @@
+ * packet will have a different size, possibly zero. Note that the encrypted packets are
+ * larger compared to the plaintext packets, because they contain a 16 byte length trailing
+ * authentication tag. The resulting encrypted and authenticated packets are assembled back into
+ * the resulting stream.
+ * <p>
+ * The packets are encrypted using the same {@link SecretKey} but using a different initialization
+ * vector. The IV is 12 bytes wide and it's comprised of an integer {@code nonce}, the same for
+ * every packet in a stream, but which MUST not otherwise be repeated for the same {@code SecretKey}
+ * across other streams, and a monotonically increasing long counter. When assembling the resulting
+ * stream, the IV is prepended to the corresponding packet's ciphertext.
+ * <p>
+ * The packet size is preferably a large multiple of the AES block size (128 bytes), but any positive
+ * integer value smaller than {@link EncryptedRepository#MAX_PACKET_LENGTH_IN_BYTES} is valid.
+ * <p>
+ * This input stream supports the {@code mark} and {@code reset} operations, but only if the wrapped
+ * stream supports them as well. A {@code mark} call will trigger the memory buffering of the current
+ * packet and will also trigger a {@code mark} call on the wrapped input stream on the next
+ * packet boundary. Upon a {@code reset} call, the buffered packet will be replayed and new packets
+ * will be generated starting from the marked packet boundary on the wrapped stream.
+ * <p>
+ * The {@code close} call will close the encryption input stream and any subsequent {@code read},
+ * {@code skip}, {@code available} and {@code reset} calls will throw {@code IOException}s.
+ * <p>
+ * This is NOT thread-safe, multiple threads sharing a single instance must synchronize access.
+ *
+ * @see DecryptionPacketsInputStream
+ */
+public final class EncryptionPacketsInputStream extends ChainingInputStream {
+
+    protected final InputStream source; // protected for tests
+    private final SecretKey secretKey;
+    private final int packetLength;
+    private final ByteBuffer packetIv;
+    private final int encryptedPacketLength;
+
+    protected long counter; // protected for tests
+    protected Long markCounter; // protected for tests
+    protected int markSourceOnNextPacket; // protected for tests
+
+    public static long getEncryptionSize(long size, int packetLength) {
+        return size + (size / packetLength + 1) * (EncryptedRepository.GCM_TAG_SIZE_IN_BYTES + EncryptedRepository.GCM_IV_SIZE_IN_BYTES);
+    }
+
+    public EncryptionPacketsInputStream(InputStream source, SecretKey secretKey, int nonce, int packetLength) {
+        this.source = Objects.requireNonNull(source);
+        this.secretKey = Objects.requireNonNull(secretKey);
+        if (packetLength <= 0 || packetLength >= EncryptedRepository.MAX_PACKET_LENGTH_IN_BYTES) {
+            throw new IllegalArgumentException("Invalid packet length [" + packetLength + "]");
```

Comment:
The former one, it should be fixed by the some higher level code, and this branch would throw only on internal errors.

But I can't exclude the possibility that we might expose this option in the future, to squeeze out performance. I can also imagine we might have an heuristic to adjust this for blobs of different sizes (eg suppose we learn that it's better to have larger packet sizes for larger blobs) or for different JVM Security Providers.

If this will be exposed to the administrator, there should be validation before the code hits this condition. But the code here still has value: because **decryption** must allocate a full packet worth of memory up front, it's susceptible to DOS if it doesn't clamp the maximum allowed size of the buffer, so this code here makes sure that **encryption** will not generate ciphertext (in the case of a botched higher-level validation) that will be rejected on decryption.

If this makes sense for you, I'll keep the `IllegalArgumentException` here.

Reply:
> If this makes sense for you, I'll keep the IllegalArgumentException here.

Yes, that's fine. It's what I expected, but wanted to double check.